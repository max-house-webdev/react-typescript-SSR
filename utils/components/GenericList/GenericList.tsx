import React from 'react';

export interface IItem {
  id: string;
  text: string;
  onClick?: (id: string) => void;
  className?: string;
  As?: 'a' | 'li' | 'button' | 'div';
  href?: string;
  type?: 'button' | 'submit' | 'reset';
  child?: React.ReactElement;
}

export interface IGenericListProps {
  list: IItem[];
}

export function GenericList({ list }: IGenericListProps) {
  const noop = () => {};
  return (
    <>
      {list.map(
        ({
          As = 'div',
          text,
          onClick = noop,
          className,
          id,
          href,
          type,
          child,
        }) => (
          <As
            className={className}
            onClick={() => onClick(id)}
            key={id}
            href={href}
            type={type}
          >
            {child}
            {text}
          </As>
        )
      )}
    </>
  );
}
