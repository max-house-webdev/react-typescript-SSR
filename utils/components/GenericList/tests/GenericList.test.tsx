import React from 'react';
import { shallow } from 'enzyme';

import { GenericList } from '../GenericList';
import { IItem, IGenericListProps } from '../GenericList';

describe('<GenericList />', () => {
  describe('by default', () => {
    const props: IGenericListProps = {
      list: [
        {
          id: 'id-0',
          text: 'text-0',
        },
        {
          id: 'id-1',
          text: 'text-1',
        },
        {
          id: 'id-2',
          text: 'text-1',
        },
      ],
    };

    const { list } = props;

    const wrapper = shallow(<GenericList list={list} />);

    test('should render list of three DivElement', () => {
      expect(wrapper.find('div')).toHaveLength(3);
      expect(wrapper.find('div#id-0')).toBeDefined();
      expect(wrapper.find('div#id-1')).toBeDefined();
      expect(wrapper.find('div#id-2')).toBeDefined();
    });

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('custom', () => {
    const props: IGenericListProps = {
      list: [
        {
          id: 'id-0',
          text: 'text-0',
          As: 'a',
          className: 'list__item',
          child: <span>child</span>,
          onClick: () => {
            return 'click 0';
          },
        },
        {
          id: 'id-1',
          text: 'text-1',
          As: 'a',
          className: 'list__item',
          child: <span>child</span>,
          onClick: () => {
            return 'click 1';
          },
        },
      ],
    };

    const { list } = props;

    const wrapper = shallow(<GenericList list={list} />);

    test('should render list of two AnchorElement', () => {
      expect(wrapper.find('a.list__item')).toHaveLength(2);
    });

    test('should contain onclick Fn', () => {
      const result = wrapper.find('a').map((listItem) => {
        return listItem.simulate('click');
      });

      expect(result).toHaveLength(2);
    });
  });
});
