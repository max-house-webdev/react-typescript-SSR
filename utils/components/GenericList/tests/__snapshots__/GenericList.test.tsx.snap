// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenericList /> by default should match snapshot 1`] = `
<Fragment>
  <div
    key="id-0"
    onClick={[Function]}
  >
    text-0
  </div>
  <div
    key="id-1"
    onClick={[Function]}
  >
    text-1
  </div>
  <div
    key="id-2"
    onClick={[Function]}
  >
    text-1
  </div>
</Fragment>
`;
